src/main/java/com/uid2/client/PublisherUid2Helper.java

```diff
@@ -0,0 +1,129 @@
+
+        if (!"success".equals(jsonIdentity.get("status").getAsString())) {
+            throw new Uid2Exception("Got unexpected token generate status in decrypted response:\n" + new GsonBuilder().setPrettyPrinting().create().toJson(jsonIdentity));
+        }
+
+        return IdentityTokens.fromJson(getBodyAsJson(jsonIdentity));
+    }
+
+    /**
+     * @param encryptedResponse the response body returned by a call to <a href="https://github.com/UnifiedID2/uid2docs/blob/main/api/v2/endpoints/post-token-refresh.md">/token/refresh</a>
+     * @param currentIdentity the current IdentityTokens instance, typically retrieved from a user's session
+     * @return the refreshed IdentityTokens instance (with a new advertising token and updated expiry times). Typically, this will be used to replace the current identity in the user's session
+     */
+    public static IdentityTokens createIdentityFromTokenRefreshResponse(String encryptedResponse, IdentityTokens currentIdentity) {
+        return createIdentityFromTokenRefreshResponseImpl(encryptedResponse, currentIdentity, Instant.now());
+    }
+
+    EnvelopeV2 createEnvelopeImpl(TokenGenerateInput tokenGenerateInput, byte[] nonce, Instant timestamp, byte[] iv) {
+        byte[] jsonEmailBytes = tokenGenerateInput.getAsJsonString().getBytes(StandardCharsets.UTF_8);
+
+        //Note this is very similar to UID2Client.makeV2Request (could share implementation)
+        ByteBuffer writer = ByteBuffer.allocate(TIMESTAMP_LENGTH + nonce.length + jsonEmailBytes.length);
+        writer.putLong(timestamp.toEpochMilli());
+        writer.put(nonce);
+        writer.put(jsonEmailBytes);
+
+        byte[] encrypted = Decryption.encryptGCM(writer.array(), iv, secretKey);
+        ByteBuffer envelopeBuffer = ByteBuffer.allocate(encrypted.length + 1);
+        final byte envelopeVersion = 1;
+        envelopeBuffer.put(envelopeVersion);
+        envelopeBuffer.put(encrypted);
+        return new EnvelopeV2(InputUtil.byteArrayToBase64(envelopeBuffer.array()), nonce);
+    }
+
+    static IdentityTokens createIdentityFromTokenRefreshResponseImpl(String encryptedResponse, IdentityTokens currentIdentity, Instant timestamp) {
+        String response;
+        String refreshResponseKey = currentIdentity.getRefreshResponseKey();
+        if (refreshResponseKey != null) {
+            response = decrypt(encryptedResponse, InputUtil.base64ToByteArray(refreshResponseKey), true, null);
+        } else { //if refresh_response_key doesn't exist, assume refresh_token came from a v1/token/generate query. In that scenario, /v2/token/refresh will return an unencrypted response.
+            response = encryptedResponse;
+        }
+
+        JsonObject responseJson = new Gson().fromJson(response, JsonObject.class);
+        String status = responseJson.get("status").getAsString();
+
+        if ("optout".equals(status)) {
+            return null;
```

Comment:
Should it be more obvious to the end user that this result is because of an opt-out? Not sure how else you could do this other than wrap the IdentityTokens or throw an exception... Just null feels a bit odd to me.

Reply:
Agree, I'm going to wrap this